man/jsemcgen.8
.\" section 8 man page for jsemcgen.sh
.\"
.\" This man page was first written by Cody Boone Ferguson for the IOCCC
.\" in 2022.
.\"
.\" Humour impairment is not virtue nor is it a vice, it's just plain
.\" wrong: almost as wrong as JSON spec mis-features and C++ obfuscation! :-)
.\"
.\" "Share and Enjoy!"
.\"     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
.\"
.TH jsemcgen.sh 8 "27 November 2022" "jsemcgen.sh" "IOCCC tools"
.SH NAME
jsemcgen.sh \- generate JSON semantics table via the \fBjsemtblgen(8)\fP tool
.SH SYNOPSIS
\fBjsemcgen.sh\fP [\-h] [\-v level] [\-J level] [\-q] [\-V] [\-s] [\-I] [\-N name] [\-D def_func] [\-P prefix] [\-1 func] [\-S func] [\-B func] [\-0 func] [\-M func] [\-O func] [\-A func] [\-U func] [\-j jsemtblgen] [\-p patch_tool] file.json head patch tail out
.SH DESCRIPTION
This tool is a wrapper to the \fBjsemtblgen(8)\fP tool. Along with \fBall_ref(8)\fP these tools create the JSON semantics tables for the \fBchkentry(1)\fP tool.
As the \fBchkentry(1)\fP tool is a vitally important part of the IOCCC judging process these tools are also vital in running the IOCCC but at a lower level.
.PP
The \fIfile.json\fP file is expected to be a valid JSON file used to generate the initial JSON semantics table.
The \fIhead\fP file is a file add to the beginning of the \fBjsemtblgen(8)\fP output (before patched \fBjsemtblgen(8)\fP output).
The \fIpatch\fP file is a patch file to apply to the output of \fBjsemtblgen(8)\fP with \fIhead\fP and \fItail\fP.
The \fItail\fP file is a file to add to the end of the \fBjsemtblgen(8)\fP output (after patching the \fBjsemtblgen(8)\fP output).
The \fIout\fP argument is the \fBjsemtblgen(8)\fP output with possible head, tail and patch applied.
If you do not wish to specify \fIhead\fP, \fIpatch\fP or \fItail\fP specify \fI.\fP instead.
.SH OPTIONS
.TP
\fB\-h\fP
Show help and exit.
.TP
\fB\-v \fIlevel\fP\fP
Set verbosity level to level.
.TP
\fB\-J \fIlevel\fP\fP
Set JSON verbosity level.
.TP
\fB\-q\fP
Set quiet mode.
.TP
\fB\-V\fP
Show version and exit.
.TP
\fB\-s\fP
Specify that arg is a string to be processed rather than a file.
.TP
\fB\-I\fP
Output as a C header file.
.TP
\fB\-N \fIname\fP\fP
Set name of the semantics table.
.TP
\fB\-D \fIdef_func\fP\fP
Validate with \fIdef_func()\fP unless overridden by another flag.
.TP
\fB\-P \fIprefix\fP\fP
Validate \fBJTYPE_MEMBER\fP JSON nodes with prefix_name() (def: do not)
.RS
Note that the name is based on the \fBJTYPE_MEMBER\fP JSON decoded name string.
Underscore (_) replaces any name chars that are not valid in a C function name.
\fB\-P\fP overrides any use of \fB\-M\fP.
.RE
.TP
\fB\-1 \fIfunc\fP\fP
Validate \fBJTYPE_NUMBER\fP JSON nodes with \fIfunc()\fP.
.TP
\fB\-S \fIfunc\fP\fP
Validate \fBJTYPE_STRING\fP JSON nodes with \fIfunc()\fP.
.TP
\fB\-B \fIfunc\fP\fP
Validate \fBJTYPE_BOOL\fP JSON nodes with \fIfunc()\fP.
.TP
\fB\-0 \fIfunc\fP\fP
Validate \fBJTYPE_NULL\fP JSON nodes with \fIfunc()\fP.
.TP
\fB\-M \fIfunc\fP\fP
Validate \fBJTYPE_MEMBER\fP JSON nodes with \fIfunc()\fP.
.TP
\fB\-O \fIfunc\fP\fP
Validate \fBJTYPE_OBJECT\fP JSON nodes with \fIfunc()\fP.
.TP
\fB\-A \fIfunc\fP\fP
Validate \fBJTYPE_ARRAY\fP JSON nodes with \fIfunc()\fP.
.TP
\fB\-U \fIfunc\fP\fP
Validate nodes with unknown types with \fIfunc()\fP.
.TP
\fB\--j \fIjsemtblgen\fP\fP
Set path to \fBjsemtblgen(8)\fP.
.RS
This option is only used in this script.
.RE
.TP
\fB\-p \fIpatch_tool\fP\fP
Set path to the patch program.
.RS
This option is only used in this script.
.RE
.SH EXIT STATUS
.TP
0
JSON is valid.
.TQ
1
JSON is invalid.
.TQ
2
\fB\-h\fP or \fB\-V\fP used and help string or version string printed.
.TQ
3
Command line error.
.TQ
4
\fBjsemtblgen\fP or \fBpatch_tool\fP is not an executable.
.TQ
5
One or more of the files \fIfile.json\fP, \fIhead\fP, \fIpatch\fP, and/or \fItail\fP is or are not a readable file.
.TQ
6
\fBjsemtblgen(8)\fP failed.
.TQ
>= 10
Internal error.
.SH NOTES
.PP
The JSON parser \fBjparse\fP was co\-developed by Cody Boone Ferguson and Landon Curt Noll (one of the IOCCC Judges) in support for IOCCCMOCK, IOCCC28 and beyond.
.PP
This tool is needed in order to generate the JSON semantics table for \fBchkentry(1)\fP.
.SH BUGS
If you have a problem with the tool (not JSON itself! :\-) ) you can report it at the GitHub issues page.
It can be found at
.br
\fI\<https://github.com/ioccc\-src/mkiocccentry/issues\>\fP.
.SH SEE ALSO
\fBjsemtblgen(8)\fP, \fBchkentry(1)\fP, \fBall_ref(8)\fP


CHANGES.md
# Major changes to the IOCCC entry toolkit


## Release 0.8.1 2022-11-19

This is "roll-up" release was added so that it could be referenced
in the https://www.ioccc.org/index.html news section.

Updated legacy_clobber rule to remove older directories no longer used.

Moved test directory to test_ioccc to avoid Makefile confusion
with the test rule.

Restructure dbg facility and Makefile in accordance with
the new method of copying code from external GitHub repositories.

Improve formatting with the hostchk_warning rule.

Added initial dyn_array.md documentation.

Typo fixes to the verge.8 man page.

Minor fixes to jparse.md.

Updated txzchk comments (source, header, man page).

Improved and updated README.md documentation.

Improved bug_report.sh: in particular in reporting versions.

Bug fixes and improvements to several shell scripts.
Made changes to satisfy shellcheck as well.

Added rules in the Makefile to make is easier to import code from
external GitHub repositories.

```
This repo is designed to be a standalone repo.  Even though we use other
repositories, we prefer to NOT clone them.  We want this repo to depend
on a specific version of such code so that a change in the code of those
external repositories will NOT impact this repo.

For that reason, and others, we maintain a private copy of an external
repository as clone.repo.  The clone.repo directory is excluded from
out repo via the .gitignore file.  We copy clone.repo/ into repo/ and
check in those file directly into this repo.
```

The dbg code is now a copy of the code from the dbg repo.

Update JSON parser version.  Changed JSON_PARSER_VERSION from "0.11
2022-11-04" to "0.12 2022-11-09".  There was an API change (an
addition of a filename argument in some calls) that suggested version change.


## Release 0.8 2022-11-09

Improve the situation with some systems requiring some feature test macros.
Removed all references to `${TIMEGM_PROBLEM}` in Makefile.
Removed all references to `${D_LEGACY}` in Makefile.
Cleaned out the `$(WARN_FLAGS}` list (from legacy hosts).
These changes allow for cleanly compiling under CentOS without as many
complexities (it was also important for modern systems like Fedora).

Made a number of important fixes to `chkentry`.

Made a number of important fixes to JSON parser.
Pre-scan documents and strings for problematic bytes prior to trying to parse it
as json.

The JSON scanner and parser are now re-entrant!

Improved JSON parser error messages by adding filename and location (lines and
columns).

Added filename argument to top level parser functions:

```c
extern struct json *parse_json(char const *ptr, size_t len, char const *filename, bool *is_valid);
extern struct json *parse_json_stream(FILE *stream, char const *filename, bool *is_valid);
```

This was one part of making the scanner and parser re-entrant. See the git log
and diffs of `jparse.l` and `jparse.y` for more details.

The documentation of the JSON parser is not complete and it will not be
completed until after IOCCCMOCK.

Fixed several memory leaks.

Fixed a number of typos in comments.

Many man pages updated and/or fixed.
Removed `limit_ioccc(8)` man page.

Made a number of shell script fixes with respect to the `mktemp(1)` command.

Experimenting with a system independent `make depend` tool: work in progress.


## Release 0.7 2022-11-05

Released IOCCC entry toolkit v0.7 2022-11-05

Changed BUG_REPORT_VERSION from "0.5 2022-11-03" to "0.6 2022-11-06".
Changed FMT_DEPEND_VERSION from "1.0 2022-09-28" to "1.1 2022-11-05".
Changed CHKENTRY_TEST_VERSION from "0.1 2022-10-11" to "0.2 2022-11-04".
Changed IOCCC_TEST_VERSION from "0462202-04-23" (sic) to "0.5 2022-11=04".
Changed TEST_IOCCCSIZE_VERSION from "1.1 2022-10-16" to ="1.2 2022-11-04".
Changed JPARSE_TEST_VERSION from "0.3 2022-10-22" to "0.4 2022-11-04".
Changed JSTR_TEST_VERSION from "0.4 2022-09-28" to "0.4 2022-11-05".
Changed TXZCHK_TEST_VERSION from "0.4 2022-10-20" to "0.5 2022-11-04".

The `chkentry(1)` tool is now code complete!  Added test suite for
`chkentry` called test/chkentry_test.sh.  The `chkentry(1)` tool
passes all tests.

Removed prefix in bison and flex code. The programmer's apology and
warning are still in `sorry.tm.ca.h`.

A test suite for `txzchk` was added.

Removed the `MAX_DIR_KSIZE` limit. There still is a maximum tarball
size but now there also is a limit of the number of files in the
tarball. `txzchk` and `mkiocccentry` were updated for this.

We're one step closer to being able to invite the public to
review these tools.

Created the test sub-directory.  Moved many test tools under the
test sub-directory.  Moved log files under test.

Created the man sub-directory.  Moved all man pages under the
man sub-directory.  The html files are also build under man.

Added the soup sub-directory.  Moved tools related to building semantic
test tables for chkentry under the soup sub-directory.  Moved ref
under soup.

Added fmt_depend.sh under the soup sub-directory to format
the make depend lines in a system independent way.

Added -Z topdir flags to soup/fmt_depend.sh, test/ioccc_test.sh,
test/jstr_test.sh, test/mkiocccentry_test.sh, and test/txzchk_test.sh.
By default these tools cd to the top level directory where the
mkiocccentry.c source file resides.  By use of -Z topdir, one can
force the top level directory location.

Experimenting with compiling code in sub-directories such as soup
and test.


## Release 0.6 2022-09-02

Released IOCCC entry toolkit v0.6 2022-09-02

Updated CHANGES.md for v0.6 2022-09-02.

Changed `MIN_TIMESTAMP` from 1655958810 to 1662145368.

Updated `mkiocccentry` from "0.40 2022-03-15" to "0.41 2022-09-02".
Updated `.info.json` version from "1.10 2022-06-22" to "1.11 2022-09-02".
Updated `.author.json` version from "1.13 2022-06-22" to "1.14 2022-09-02".

Improved code to use new facilities for output to a buffer from dbg release of
v2.5 2022-07-23.


jsemcgen.sh
#!/usr/bin/env bash
#
# jsemcgen.sh - generate JSON semantics table
#
# The concept of this file was developed by:
#
#	chongo (Landon Curt Noll, http://www.isthe.com/chongo/index.html) /\oo/\
#
# The JSON parser was co-developed by:
#
#	@xexyl
#	https://xexyl.net		Cody Boone Ferguson
#	https://ioccc.xexyl.net
# and:
#	chongo (Landon Curt Noll, http://www.isthe.com/chongo/index.html) /\oo/\
#
# Share and enjoy! :-)
#     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
#

# setup
#
export V_FLAG="0"
export J_FLAG="0"
export Q_FLAG=
export S_FLAG=
export I_FLAG=
export NAME=
export DEFAULT_FUNC=
export PREFIX=
export NUMBER_FUNC=
export STRING_FUNC=
export BOOL_FUNC=
export NULL_FUNC=
export MEMBER_FUNC=
export OBJECT_FUNC=
export ARRAY_FUNC=
export UNKNOWN_FUNC=
export JSEMTBLGEN="./jsemtblgen"
export JSEMTBLGEN_ARGS=
export PATCH_TOOL=
export JSEMCGEN_VERSION="1.2 2022-10-23"

# attempt to fetch system specific paths to tools we need
#
# Try to find the patch tool.
#
PATCH_TOOL="$(type -P patch 2>/dev/null)"
# It's possible that the path could not be obtained so we set it to the default
# in this case.
#
# We cannot do it via parameter substitution because patch without an arg is not
# an error (it tries to execute it) so we do it via a direct check to the
# variable being empty. If it would work it would look like:
#
#   ${PATCH_TOOL:=/usr/bin/patch} 2>/dev/null
#

# make sure PATCH_TOOL is set
#
if [[ -z "$PATCH_TOOL" ]]; then
    PATCH_TOOL="/usr/bin/patch"
fi

export USAGE="usage: $0 [-h] [-v level] [-J level] [-q] [-V] [-s] [-I] [-N name] [-D def_func] [-P prefix]
	[-1 func] [-S func] [-B func] [-0 func] [-M func] [-O func] [-A func] [-U func]
	[-j jsemtblgen] [-p patch_tool] file.json head patch tail out

	-h		print help message and exit
	-v level	set verbosity level (def level: 0)
	-J level	set JSON verbosity level (def level: 0)
	-q		quiet mode: silence msg(), warn(), warnp() if -v 0 (def: not quiet)
	-V		print version string and exit
	-s		file.json is a string (def: arg is a filename)

	-I		output as .h include file (def: output as .c src)

	-N name		name of the semantics table (def: sem_tbl)

	-D def_func	validate with def_func() unless overridden by another flag (def: NULL)

	-P prefix	validate JTYPE_MEMBER JSON nodes with prefix_name() (def: do not)

			NOTE: The name is based on the JTYPE_MEMBER JSON decoded name string.
			NOTE: Underscore (_) replaces any name chars that are not valid in a C function name.
			NOTE: -P overrides any use of -M.

	-1 func		validate JTYPE_NUMBER JSON nodes with func() (def: do not)
	-S func		validate JTYPE_STRING JSON nodes with func() (def: do not)
	-B func		validate JTYPE_BOOL JSON nodes with func() (def: do not)
	-0 func		validate JTYPE_NULL JSON nodes with func() (def: do not)
	-M func		validate JTYPE_MEMBER JSON nodes with func() (def: do not)
	-O func		validate JTYPE_OBJECT JSON nodes with func() (def: do not)
	-A func		validate JTYPE_ARRAY JSON nodes with func() (def: do not)
	-U func		validate nodes with unknown types with func() (def: do not)

	NOTE: All of the above are passed down to the jsemtblgen tool.
	      All of the below are only used by this script.

	-j jsemtblgen	path to jsemtblgen (def: ./jsemtblgen)
	-p patch_tool	path to the patch tool (def: $PATCH_TOOL)

	file.json	A valid JSON file used to generate the initial JSON semantics table
	head		file add to the beginning of the jsemtblgen output (before patched jsemtblgen output), . ==> do not add
	patch		patch to apply to the output of jsemtblgen with head and tail, . --> do not patch
	tail		file add to the end of jsemtblgen output (after patched jsemtblgen output), . ==> do not add
	out		jsemtblgen output, with possible head and tail and patch applied

Exit codes:
     0	 JSON is valid
     1	 JSON is invalid
     2	 -h and help string printed or -V and version string printed
     3	 command line error
     4	 jsemtblgen or patch_tool is not an executable
     5	 file.json, head, patch, and/or tail is not a readable file
     6	 jsemtblgen failed
 >= 10	 internal error"

# parse args
#
while getopts :hv:J:qVsIN:D:P:1:S:B:0:M:O:A:U:j:p: flag; do
    case "$flag" in
    h) echo "$USAGE" 1>&2
       exit 2
       ;;
    v) V_FLAG="$OPTARG";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -v '$V_FLAG'";
       ;;
    J) J_FLAG="$OPTARG";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -J '$J_FLAG'";
       ;;
    q) Q_FLAG="-q";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -q";
       ;;
    V) echo "$JSEMCGEN_VERSION"
       exit 2
       ;;
    s) S_FLAG="-s";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -s";
       ;;
    I) I_FLAG="-I";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -I";
       ;;
    N) NAME="$OPTARG";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -N '$NAME'";
       ;;
    D) DEFAULT_FUNC="$OPTARG";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -D '$DEFAULT_FUNC'";
       ;;
    P) PREFIX="$OPTARG";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -P '$PREFIX'";
       ;;
    1) NUMBER_FUNC="$OPTARG";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -1 '$NUMBER_FUNC'";
       ;;
    S) STRING_FUNC="$OPTARG";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -S '$STRING_FUNC'";
       ;;
    B) BOOL_FUNC="$OPTARG";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -B '$BOOL_FUNC'";
       ;;
    0) NULL_FUNC="$OPTARG";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -0 '$NULL_FUNC'";
       ;;
    M) MEMBER_FUNC="$OPTARG";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -M '$MEMBER_FUNC'";
       ;;
    O) OBJECT_FUNC="$OPTARG";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -O '$OBJECT_FUNC'";
       ;;
    A) ARRAY_FUNC="$OPTARG";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -A '$ARRAY_FUNC'";
       ;;
    U) UNKNOWN_FUNC="$OPTARG";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -U '$UNKNOWN_FUNC'";
       ;;
    j) JSEMTBLGEN="$OPTARG";
       ;;
    p) PATCH_TOOL="$OPTARG";
       ;;
    \?) echo "$0: ERROR: invalid option: -$OPTARG" 1>&2
       exit 3
       ;;
    :) echo "$0: ERROR: option -$OPTARG requires an argument" 1>&2
       exit 3
       ;;
   *)
       ;;
    esac
done

# check args
#
shift $(( OPTIND - 1 ));
if [[ $# -ne 5 ]]; then
    echo "$0: ERROR: expected 5 arguments, found $#" 1>&2
    echo "$USAGE" 1>&2
    exit 3
fi
JSON_FILE="$1"
JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -- '$JSON_FILE'"
HEAD_FILE="$2"
PATCH_FILE="$3"
TAIL_FILE="$4"
OUT_FILE="$5"
export JSON_FILE HEAD_FILE PATCH_FILE TAIL_FILE OUT_FILE
if [[ $HEAD_FILE = "." ]]; then
    HEAD_FILE=
fi
if [[ $PATCH_FILE = "." ]]; then
    PATCH_FILE=
fi
if [[ $TAIL_FILE = "." ]]; then
    TAIL_FILE=
fi
if [[ $V_FLAG -ge 5 ]]; then
    echo "$0: debug[5]: V_FLAG: $V_FLAG" 1>&2
    echo "$0: debug[5]: J_FLAG: $J_FLAG" 1>&2
    echo "$0: debug[5]: Q_FLAG: $Q_FLAG" 1>&2
    echo "$0: debug[5]: S_FLAG: $S_FLAG" 1>&2
    echo "$0: debug[5]: I_FLAG: $I_FLAG" 1>&2
    echo "$0: debug[5]: NAME: $NAME" 1>&2
    echo "$0: debug[5]: DEFAULT_FUNC: $DEFAULT_FUNC" 1>&2
    echo "$0: debug[5]: PREFIX: $PREFIX" 1>&2
    echo "$0: debug[5]: NUMBER_FUNC: $NUMBER_FUNC" 1>&2
    echo "$0: debug[5]: STRING_FUNC: $STRING_FUNC" 1>&2
    echo "$0: debug[5]: BOOL_FUNC: $BOOL_FUNC" 1>&2
    echo "$0: debug[5]: NULL_FUNC: $NULL_FUNC" 1>&2
    echo "$0: debug[5]: MEMBER_FUNC: $MEMBER_FUNC" 1>&2
    echo "$0: debug[5]: OBJECT_FUNC: $OBJECT_FUNC" 1>&2
    echo "$0: debug[5]: ARRAY_FUNC: $ARRAY_FUNC" 1>&2
    echo "$0: debug[5]: UNKNOWN_FUNC: $UNKNOWN_FUNC" 1>&2
    echo "$0: debug[5]: jsemtblgen: $JSEMTBLGEN" 1>&2
    echo "$0: debug[5]: jsemtblgen args: $JSEMTBLGEN_ARGS" 1>&2
    echo "$0: debug[5]: file.json file: $JSON_FILE" 1>&2
    if [[ -z $HEAD_FILE ]]; then
	echo "$0: debug[5]: no head file" 1>&2
    else
	echo "$0: debug[5]: head file: $HEAD_FILE" 1>&2
    fi
    if [[ -z $PATCH_FILE ]]; then
	echo "$0: debug[5]: no patch file" 1>&2
    else
	echo "$0: debug[5]: patch file: $PATCH_FILE" 1>&2
    fi
    if [[ -z $TAIL_FILE ]]; then
	echo "$0: debug[5]: no tail file" 1>&2
    else
	echo "$0: debug[5]: tail file: $TAIL_FILE" 1>&2
    fi
    echo "$0: debug[5]: out file: $OUT_FILE" 1>&2
fi

# firewall
#
if [[ ! -x $JSEMTBLGEN ]]; then
    echo "$0: ERROR: jsemtblgen is not an executable: $JSEMTBLGEN" 1>&2
    exit 4
fi
if [[ -z $PATCH_TOOL ]]; then
    echo "$0: ERROR: cannot find default patch tool path, need to use -p patch_path" 1>&2
    exit 4
fi
if [[ ! -x $PATCH_TOOL ]]; then
    echo "$0: ERROR: patch tool is not an executable: $PATCH_TOOL" 1>&2
    exit 4
fi
if [[ ! -e $JSON_FILE ]]; then
    echo "$0: ERROR: file.json does not exist: $JSON_FILE" 1>&2
    exit 5
fi
if [[ ! -f $JSON_FILE ]]; then
    echo "$0: ERROR: file.json is not a file: $JSON_FILE" 1>&2
    exit 5
fi
if [[ ! -r $JSON_FILE ]]; then
    echo "$0: ERROR: file.json is not a readable file: $JSON_FILE" 1>&2
    exit 5
fi
if [[ -n $HEAD_FILE ]]; then
    if [[ ! -e $HEAD_FILE ]]; then
	echo "$0: ERROR: head does not exist: $HEAD_FILE" 1>&2
	exit 5
    fi
    if [[ ! -f $HEAD_FILE ]]; then
	echo "$0: ERROR: head is not a file: $HEAD_FILE" 1>&2
	exit 5
    fi
    if [[ ! -r $HEAD_FILE ]]; then
	echo "$0: ERROR: head is not a readable file: $HEAD_FILE" 1>&2
	exit 5
    fi
fi
if [[ -n $PATCH_FILE ]]; then
    if [[ ! -e $PATCH_FILE ]]; then
	echo "$0: ERROR: patch does not exist: $HEAD_FILE" 1>&2
	exit 5
    fi
    if [[ ! -f $PATCH_FILE ]]; then
	echo "$0: ERROR: patch is not a file: $PATCH_FILE" 1>&2
	exit 5
    fi
    if [[ ! -r $PATCH_FILE ]]; then
	echo "$0: ERROR: patch is not a readable file: $PATCH_FILE" 1>&2
	exit 5
    fi
fi
if [[ -n $TAIL_FILE ]]; then
    if [[ ! -e $TAIL_FILE ]]; then
	echo "$0: ERROR: tail does not exist: $TAIL_FILE" 1>&2
	exit 5
    fi
    if [[ ! -f $TAIL_FILE ]]; then
	echo "$0: ERROR: tail is not a file: $TAIL_FILE" 1>&2
	exit 5
    fi
    if [[ ! -r $TAIL_FILE ]]; then
	echo "$0: ERROR: tail is not a readable file: $TAIL_FILE" 1>&2
	exit 5
    fi
fi

# case: jsemtblgen is patched
#
if [[ -n "$PATCH_FILE" ]]; then

    # obtain a temporary filenames
    #
    export MKTEMP_TEMPLATE TMP_FILE
    MKTEMP_TEMPLATE=".jsemcgen.XXXXXXXXXX.out"
    if [[ $V_FLAG -ge 3 ]]; then
	echo "$0: debug[3]: about to run: mktemp $MKTEMP_TEMPLATE" 1>&2
    fi
    TMP_FILE=$(mktemp "$MKTEMP_TEMPLATE")
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: mktemp $MKTEMP_TEMPLATE exit code: $status" 1>&2
	exit 10
    fi
    if [[ ! -e $TMP_FILE ]]; then
	echo "$0: ERROR: tmp file not found: $TMP_FILE" 1>&2
	exit 11
    fi
    if [[ ! -f $TMP_FILE ]]; then
	echo "$0: ERROR: tmp not a regular file: $TMP_FILE" 1>&2
	exit 12
    fi
    if [[ ! -r $TMP_FILE ]]; then
	echo "$0: ERROR: tmp not a readable file: $TMP_FILE" 1>&2
	exit 13
    fi
    if [[ ! -w $TMP_FILE ]]; then
	echo "$0: ERROR: tmp not a writable file: $TMP_FILE" 1>&2
	exit 14
    fi
    if [[ $V_FLAG -ge 3 ]]; then
	trap "exit" 0 1 2 3 15
    else
	trap "rm -f \$TMP_FILE \$TMP_FILE.rej \$TMP_FILE.orig; exit" 0 1 2 3 15
    fi
    export ORIG_FILE="$TMP_FILE.orig"
    export REJ_FILE="$TMP_FILE.rej"
    export PATH_ERR="$TMP_FILE.err"

    # output the header
    #
    if [[ -n "$HEAD_FILE" ]]; then
	if [[ $V_FLAG -ge 1 ]]; then
	    echo "$0: debug[1]: cat $HEAD_FILE > $TMP_FILE"
	fi
	cat "$HEAD_FILE" > "$TMP_FILE"
	status="$?"
	if [[ $status -ne 0 ]]; then
	    echo "$0: ERROR: cat $HEAD_FILE > $TMP_FILE exit status: $status" 1>&2
	    exit 15
	fi
    fi

    # write the jsemtblgen generated code to a temporary file
    #
    if [[ $V_FLAG -ge 1 ]]; then
	echo "$0: debug[1]: about to run: $JSEMTBLGEN $JSEMTBLGEN_ARGS > $TMP_FILE" 1>&2
    fi
    eval "$JSEMTBLGEN" "$JSEMTBLGEN_ARGS" >> "$TMP_FILE"
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: jsemtblgen failed, exit status: $status" 1>&2
	exit 16
    fi
    if [[ ! -e $TMP_FILE ]]; then
	echo "$0: ERROR: jsemtblgen tmp file does not exist: $TMP_FILE" 1>&2
	exit 17
    fi
    if [[ ! -f $TMP_FILE ]]; then
	echo "$0: ERROR: jsemtblgen tmp file is not a file: $TMP_FILE" 1>&2
	exit 18
    fi
    if [[ ! -r $TMP_FILE ]]; then
	echo "$0: ERROR: jsemtblgen tmp file is not a readable file: $TMP_FILE" 1>&2
	exit 19
    fi
    if [[ ! -w $TMP_FILE ]]; then
	echo "$0: ERROR: jsemtblgen tmp file is not a writable file: $TMP_FILE" 1>&2
	exit 20
    fi

    # output the trailer
    #
    if [[ -n "$TAIL_FILE" ]]; then
	if [[ $V_FLAG -ge 1 ]]; then
	    echo "$0: debug[1]: cat $TAIL_FILE > $TMP_FILE"
	fi
	cat "$TAIL_FILE" >> "$TMP_FILE"
	status="$?"
	if [[ $status -ne 0 ]]; then
	    echo "$0: ERROR: cat $TAIL_FILE >> $TMP_FILE exit status: $status" 1>&2
	    exit 21
	fi
    fi

    # patch the temporary file
    #
    if [[ $V_FLAG -ge 3 ]]; then
	echo "$0: debug[3]: about to run: rm -f $ORIG_FILE $REJ_FILE" 1>&2
    fi
    rm -f "$ORIG_FILE" "$REJ_FILE"
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: rm -f $ORIG_FILE $REJ_FILE filed, exit status: $status" 1>&2
	exit 22
    fi
    if [[ $V_FLAG -ge 3 ]]; then
	echo "$0: debug[3]: about to run: $PATCH_TOOL -s -u -f -- $TMP_FILE $PATCH_FILE" 1>&2
    fi
    "$PATCH_TOOL" -s -u -f -- "$TMP_FILE" "$PATCH_FILE"
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: patch failed, exit status: $status" 1>&2
	exit 23
    fi
    if [[ -s $REJ_FILE ]]; then
	echo "$0: ERROR: patch did not succeed, reject file found: $REJ_FILE" 1>&2
	exit 24
    fi

    # move patched jsemtblgen generated code to output file
    #
    mv -f "$TMP_FILE" "$OUT_FILE"
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: mv -f $TMP_FILE $OUT_FILE exit status: $status" 1>&2
	exit 25
    fi

    # cleanup
    #
    if [[ $V_FLAG -ge 3 ]]; then
	echo "$0: debug[3]: preserving possible patch files: $TMP_FILE $ORIG_FILE $REJ_FILE" 1>&2
    else
	if [[ $V_FLAG -ge 1 ]]; then
	    echo "$0: debug[3]: about to run: rm -f $ORIG_FILE $REJ_FILE" 1>&2
	fi
	rm -f "$TMP_FILE" "$ORIG_FILE" "$REJ_FILE"
	status="$?"
	if [[ $status -ne 0 ]]; then
	    echo "$0: ERROR: rm -f $TMP_FILE $ORIG_FILE $REJ_FILE filed, exit status: $status" 1>&2
	    exit 26
	fi
    fi
    trap - 0 1 2 3 15

# case: jsemtblgen is not patched
#
else

    # output the header
    #
    if [[ -n "$HEAD_FILE" ]]; then
	if [[ $V_FLAG -ge 1 ]]; then
	    echo "$0: debug[1]: about to run: cat $HEAD_FILE > $OUT_FILE" 1>&2
	fi
	cat "$HEAD_FILE" > "$OUT_FILE"
	status="$?"
	if [[ $status -ne 0 ]]; then
	    echo "$0: ERROR: cat $HEAD_FILE > $OUT_FILE exit status: $status" 1>&2
	    exit 27
	fi
    else
	if [[ $V_FLAG -ge 1 ]]; then
	    echo "$0: debug[1]: about to run: rm -f $OUT_FILE" 1>&2
	fi
	rm -f "$OUT_FILE"
	status="$?"
	if [[ $status -ne 0 ]]; then
	    echo "$0: ERROR: rm -f $OUT_FILE exit status: $status" 1>&2
	    exit 28
	fi
    fi

    # just output unpatched jsemtblgen generated code
    #
    if [[ $V_FLAG -ge 1 ]]; then
	echo "$0: debug[1]: about to run: $JSEMTBLGEN $JSEMTBLGEN_ARGS >> $OUT_FILE" 1>&2
    fi
    eval "$JSEMTBLGEN" "$JSEMTBLGEN_ARGS" >> "$OUT_FILE"
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: $JSEMTBLGEN $JSEMTBLGEN_ARGS >> $OUT_FILE  failed, exit status: $status" 1>&2
	exit 29
    fi

    # output the trailer
    #
    if [[ -n "$TAIL_FILE" ]]; then
	if [[ $V_FLAG -ge 1 ]]; then
	    echo "$0: debug[1]: about to run: cat $TAIL_FILE >> $OUT_FILE" 1>&2
	fi
	cat "$TAIL_FILE" >> "$OUT_FILE"
	status="$?"
	if [[ $status -ne 0 ]]; then
	    echo "$0: ERROR: cat $TAIL_FILE >> $OUT_FILE exit status: $status" 1>&2
	    exit 30
	fi
    fi
fi

# All Done!!! -- Jessica Noll, Age 2
#
exit 0
